### PR TITLE
[SPARK-22446][SQL][ML] Declare StringIndexerModel indexer udf as nondeterministic

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

UDFs that can cause runtime exception on invalid data are not safe to pushdown, because its behavior depends on its position in the query plan. Pushdown of it will risk to change its original behavior.

The example reported in the JIRA and taken as test case shows this issue. We should declare UDFs that can cause runtime exception on invalid data as non-determinstic.

This updates the document of `deterministic` property in `Expression` and states clearly an UDF that can cause runtime exception on some specific input, should be declared as non-determinstic.

## How was this patch tested?

Added test. Manually test.

### CODE REVIEW

**Comment by viirya** [comment]
We may have other udfs in other ml models that have the same issue. If this approach is ok for the reviewers, we can make those udfs as non-deterministic then.

**Comment by viirya** [comment]
Although this touches ML codes and test, it is basically related to SQL stuff (udf). So cc @cloud-fan  for review.

**Comment by WeichenXu123** [comment]
Looks reasonable, have you check other places which have similar issue ?

**Comment by viirya** [comment]
I remember there are few places with similar udf. If this is ok, I will modify other places together.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala [hunk 1]
```diff
@@ -75,6 +75,7 @@ abstract class Expression extends TreeNode[Expression] {
    * - it relies on some mutable internal state, or
    * - it relies on some implicit input that is not part of the children expression list.
    * - it has non-deterministic child or children.
+   * - it is an UDF that can cause runtime exception on some specific input.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: how about `it assumes the input satisfies some certain condition via the child operator`?

**Comment by cloud-fan** [comment]
LGTM

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala (hunk 1 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala]: Ok.

**Comment by viirya** [comment]
@WeichenXu123 I did a scan. Currently I only found `VectorAssembler`'s udf may have similar issue. Fixed and added test for it too.

File: mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala [hunk 2]
```diff
@@ -126,4 +126,25 @@ class VectorAssemblerSuite
       .setOutputCol("myOutputCol")
     testDefaultReadWrite(t)
   }
+
+  test("VectorAssembler's UDF should not apply on filtered data") {
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala]: mark the [SPARK-22446] on the test name.

**Review by WeichenXu123 — COMMENTED** [review]
LGTM if you check everywhere for this issue.
Thanks very much!

File: mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala (hunk 2 above)
Comment (viirya) [reply mllib/src/test/scala/org/apache/spark/ml/feature/VectorAssemblerSuite.scala]: Ok.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**